### ISSUE TITLE
Specify SCION-IP Gateways - outline

### ISSUE THREAD

**Issue by nicorusti** [scionassociation/scion-dp_I-D#40]
> Harald Alvestrand (meeting notes 26.07.2024)
Document SCION-IP gateways: If the goal is to document existing deployment, the gateways must be specified. This could be an additional ISE draft. The overall SCION draft “package” must have enough information to make a connection over scion

- [x] Prepare a description of the SIG (try in terms of functionality rather than how they work)
- [x] Send text to Eliot
- [x] Decide if this needs to be a separate draft (or if it can be a paragraph in one of the other drafts) --> No

#### Feedback received:
> Harald Alvestrand 
The payload of a packet can be anything (does not affect SCION operation), but the header contains an integer saying which content it is, with the defined values being “UDP”, “TCP”, and some numbers for internal use. The process for constructing an IP packet from the SCION packet (or vice versa) is not described, but presumably involves mangling checksums - -dataplane describes a “virtual header” for constructing the checksum of things that require it, which implies that a gateway to non-SCION IP would mangle checksums inside the UDP or TCP packet.
(There are references in -overview to docs describing gateways. However, I can’t see them at the moment.)

>   My response: 
Yes, we don’t yet have a draft describing gateways in detail. There was [a presentation ](https://datatracker.ietf.org/meeting/118/materials/slides-118-panrg-operational-aspects-of-scion-00)by my co-author Sam Hitz at PANRG at 119 about it. Essentially gateways run an encapsulation protocol and exchange IP routes among each other. They use a custom protocol for that, so we felt for now it was best to leave this part out and treat gateways as a normal SCION endpoint. 
This checksums is used, for example, by UDP on top of SCION. Gateways encapsulate the whole IP packet, so they don’t mangle UDP/TCP checksums in them. However, the gateway encapsulation protocol itself runs on top of UDP on top of SCION, therefore there is a ckechsum in the encapsulating UDP packet, and this uses the “virtual header”.

> Lunch discussion
Document SCION-IP gateways: If the goal is to document existing deployment, the gateways must be specified. This could be an additional ISE draft. The overall SCION draft “package”. The  must have enough information to make a connection over scion

#### Potential sources: 
- [Anapaya Docs](https://docs.anapaya.net/en/latest/edge/configuration/scion-tunneling/#edge-tunneling-sgrp) for SGRP, tunneling, also called "Prefix discovery" in open source
- [SCIONproto](https://docs.scion.org/en/latest/sig.html) for SIG framing 
- Scion book section  13.3 The SCION–IP Gateway (SIG) 
- [component analysis draft](https://www.ietf.org/archive/id/draft-rustignoli-panrg-scion-components-03.txt) section 3.1
- https://github.com/scionproto/scion/issues/2948

**Comment by nicorusti** [comment]
> On 7 Sep 2024, at 02:16, Harald Alvestrand <harald@alvestrand.no> wrote:

> I still don't think you have enough info on the IP gateway.

> In particular - does the IP gateway pair (entry + exit) act like a NAT box, matching the source address in the source domain to the gateway address in the destination domain, or is it an one-way tunnel of unchanged bits?

> In both cases - where are the IP addresses (or equivalent) carried through the tunnel? As part of the SCION address (as the presence of those forms of address in the endpoint scheme seems to indicate), or as part of the packet?

> And if the addresses are mapped, who takes care of the payload modifications that any NAT needs to do?

> Harald
(I still haven't gotten around to reading the -controlplane and -pki drafts,  but I have not forgotten them...)

**Comment by nicorusti** [comment]
Feedback Eliot Lear call 3.10.2024

SIG Description must clarify:
*  how are packets directed to the GW?
* How does GW know where to tunnel? 
* What happens at egress?
* (mention multicast left as future work). 

Show fully flashed out config of how to have SCION nodes, a device in the middle, they did this with M-BONE. And the 6 bone to play with IPv6. If you have someone who wants to build an implementation, how can they test that it works and that they can interoperate. 
What do you have to do if you want to want to add a GW. Show a reference to what you can do (guide how to participate) e.g. to SCIONLab, not docs.scion.org. Goal; meet interop requirement. No Anapaya docs (but could be). Beginning as vendor implementation, intent is to offer to community for further improvement and potential standardization later on.

**Comment by nicorusti** [comment]
For the records, the latest and greatest description for the SIG is in this commit: https://github.com/scionassociation/scion-dp_I-D/commit/148adfb6229f252844a3aa567c82e1ed8da1cbf6 . The text is below. Following a discussion with Eliot, we decided to leave this outside of the scope of this draft.

```
# Interoperability: SCION IP Gateway {#sig}

The SCION IP Gateway (SIG) enables IP packets to be tunneled over SCION to support communication between hosts that do not run a SCION implementation. A SIG acts as a router from the perspective of IP, whilst acting as SCION endpoint from the perspective of the SCION network. It is typically deployed inside the same AS-internal network as its non-SCION hosts, or at the edge of an enterprise network.

Tunneling IP traffic over SCION requires a pair of SIGs and it involves the following steps:

1. A sender sends an IP packet towards an IP destination.

2. The IP packet reaches a SIG in the sender’s network via standard IP routing.

3. Based on the destination IP address, the source (ingress) SIG determines the destination (egress) SIG's ISD-AS endpoint address. To achieve this, SIGs may be pre-configured with a static IP prefix to remote SIG SCION address mappings. Alternatively, multiple SIGs may be configured to speak a dynamic routing protocol between each other. The choice of protocol is left up to implementors and is outside of the scope of this document. Whether an egress SIG accepts or drops traffic from a given ingress SIG is up to the policies of the receiver's operator. The control mechanisms are implementation defined.

4. The ingress SIG encapsulates the original IP packet within a SCION packet and sends it to the egress SIG. If necessary, the ingress SIG performs SCION path lookups and selects a SCION path to the egress SIG.

5. The egress SIG receives the SCION packet and decapsulates the original IP packet. It then forwards the packet to the final IP destination using standard IP routing.

## SIG Framing

IP packets are encapsulated over SCION/UDP into SIG frames. Whilst in principle, a given pair of SIGs may use any tunneling protocol, existing deployments use the SIG framing as described here. This protocol is designed to:

- provide independence from the underlying SCION path MTU which can increase and decrease over time.
- provide fast detection of packet loss and subsequent recovery of decapsulation for packets that weren't lost.
- support for multiple streams within a framing session such that independent packet sequences be tunneled in parallel.

There may be multiple IP packets in a single SIG frame, and a single IP packet may be split into multiple SIG frames.

The ingress SIG initiates unidirectional packet flows with the egress SIG simply by sending the corresponding SIG frames. There is no handshake. The egress SIG, should it accept the traffic, instanciates the necessary resources on-demand to process each flow. Each such flow forms an independent sequence of packets (a stream) ordered by an incrementing sequence number. Between a given SIG ingress/egress pair, a (session ID, stream ID) pair uniquely identifies a stream.

To preserve performance, IP packets encapsulated in a single stream SHOULD leave the egress SIG in the order in which they entered it. To that end:

- The ingress SIG SHOULD encapsulate IP packets that cannot be proven independent (e.g., with the same  IP 6-tuple) in the same stream.
- The ingress SIG SHOULD encapsulate IP packets to a given stream in the order in which they were received.
- The ingress SIG SHOULD send all frames of a given stream over the same SCION path.
- The egress SIG SHOULD reassemble and forward packets from each stream, ordered by frame sequence number and packet within each frame.

The session ID part of the (session ID, stream ID) pair is used to indicate traffic priority grouping. The egress SIG MAY dedicate processing resources to each session rather than to each individual stream.

~~~~

+-----------------------+
|         SCION         |
+-----------------------+
|          UDP          |
+-----------------------+
|    SIG frame header   |
+-----------------------+
|   SIG frame payload   |
+-----------------------+

~~~~
{: #figure-28 title="SIG framing within a SCION packet"}

## SIG Frame Header

~~~~

 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|     Version   |  Session ID   |            Index              |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|         Reserved        |             Stream ID               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                       Sequence Number                         +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

~~~~
{: #figure-29 title="SIG Frame Header format"}


All fields within SIG Frame Header are in network byte order.

- `Version` (8 bits) indicates the SIG framing version. It MUST be set to zero if following this specification.
- `Session ID` (8 bits) identifies a tunneling session between a pair of SIGs.
- `Index` (16 bits) is the byte offset of the first beginning of an IP packet within the payload. If no IP packet starts in the payload, e.g. if the frame contains only the middle or trailing part of an IP packet, the field MUST be set to 0xFFFF.
- `Reserved` (12 bits): it MUST be set to zero.
- `Stream ID` (20 bits), along with the session, it identifies a unique sequence of SIG frames. Frames from the same stream are, on the egress SIG, put into the same reassembly queue. There may be multiple streams per session.
- `Sequence Number` (64 bits) indicates the position of the frame within a stream. Consecutive frames of a given stream have consecutive sequence numbers. IP packets split among multiple frames are re-assembled by concatenating the payloads of consecutive frames.

A SIG MAY drop frames. Buffering frames received out-of-order by the egress SIG is optional. The egress SIG SHOULD drop frames from a stream if unable to perform the sequence re-assembly.

The Session ID and Stream ID are chosen by the sender but the tuple MUST be unique within a session.


## SIG Frame Payload

The SIG frame payload may contain multiple IPv4 or IPv6 packets, or parts thereof. No other types of packets can be encapsulated and the packets are placed directly after one another with no padding. Handling of multicast is not covered by this specification and it is left as future work.

SIG uses the IPv4 or IPv6 'Payload Length Field to determine the size of the packet. To make the processing easier, it is REQUIRED that the fixed part of the IP header is in the frame where the IP packet begins. In other words, the initial fragment of an IPv4 packet must be at least 20 bytes long, whilst the initial fragment of an IPv6 packet must be at least 40 bytes long.

```
